sql/src/main/java/org/apache/druid/sql/calcite/planner/querygen/DruidQueryGenerator.java

```diff
@@ -58,32 +60,64 @@ public DruidQueryGenerator(PlannerContext plannerContext, DruidLogicalNode relRo
     this.vertexFactory = new PDQVertexFactory(plannerContext, rexBuilder);
   }
 
+  static class DruidNodeStack extends Stack<DruidLogicalNode>
```

Comment:
Another option instead of maintaining another parallel Stack would be to add a mutable `operandIndex` field to `DruidLogicalNode` and walk the logical nodes assigning ids once before doing the translation.

I tend to think that mutating the object to have an id and then referencing that id is a bit less error prone than the parallel Stack approach, is there a reason that it cannot work?

Reply:
I've tried different approaches; `DruidLogicalNode` -s are immutable - and supposed to describe a subtree - the index of that node in its parent is an external detail - so I don't feel it right to add it.

I've covered 2 stacks with `DruidNodeStack` - which was better as it made possible to introduce methods which could made the logic more readable.

And as alternate approach to attach the data more naturally I've added an `Entry` class which contains the `index` and the `DruidLogicalNode` - I've retained the custom methods...but it did disturbed some existing places a bit.